### PR TITLE
Bunch of fixes for Rot3

### PR DESCRIPTION
**Pull request by varunagrawal**
- Ensure axis in `AxisAngle` constructor is a unit vector. Added test.
- Improved rotation inverse with support for quaternions.
- Use `Eigen::Transpose` for transpose return type.
- Roll/Pitch/Yaw is more efficient.
- Fix/Remove old TODOs.

---
This change is [Reviewable](https://reviewable.io/reviews/borglab/gtsam/251)

### CODE REVIEW

File: gtsam/geometry/Point3.cpp [hunk 1]
```diff
@@ -29,7 +29,10 @@ bool Point3::equals(const Point3 &q, double tol) const {
 }
 
 void Point3::print(const string& s) const {
-  cout << s << *this << endl;
+  if (s.size() != 0) {
+    cout << s << " ";
```
Comment (dellaert) [inline gtsam/geometry/Point3.cpp]: Changes behavior, should revert. People might be relying on old behavior.

File: gtsam/geometry/Rot3.h [hunk 2]
```diff
@@ -194,22 +194,24 @@ namespace gtsam {
...
     /**
      * Convert from axis/angle representation
-     * @param  axisw is the rotation axis, unit length
-     * @param   angle rotation angle
+     * @param  axis is the rotation axis, unit length
+     * @param  angle rotation angle
      * @return incremental rotation
      */
-    static Rot3 AxisAngle(const Point3& axis, double angle) {
+    static Rot3 AxisAngle(const Vector3& axis, double angle) {
```
Comment (dellaert) [inline gtsam/geometry/Rot3.h]: Changing type: not a good idea: Point3 and Vector3 are not always the same. See Compilation flags.

File: gtsam/geometry/Rot3.h [hunk 3]
```diff
@@ -268,9 +270,13 @@ namespace gtsam {
     /// Syntatic sugar for composing two rotations
     Rot3 operator*(const Rot3& R2) const;
 
-    /// inverse of a rotation, TODO should be different for M/Q
+    /// inverse of a rotation
     Rot3 inverse() const {
-      return Rot3(Matrix3(transpose()));
+#ifdef GTSAM_USE_QUATERNIONS
+      return quaternion_.inverse();
```
Comment (dellaert) [inline gtsam/geometry/Rot3.h]: be explicit about calling Rot3 constructor. Should be made explicit probably.

File: gtsam/geometry/Rot3.h [hunk 4]
```diff
@@ -268,9 +270,13 @@ namespace gtsam {
...
+#else
+      return Rot3(transpose());
```
Comment (dellaert) [inline gtsam/geometry/Rot3.h]: Since you're changing it, why not Rot3(rot_.transpose()) ?

File: gtsam/geometry/Rot3.h [hunk 5]
```diff
@@ -405,8 +411,7 @@ namespace gtsam {
     /**
      * Return 3*3 transpose (inverse) rotation matrix
      */
-    Matrix3 transpose() const;
-    // TODO: const Eigen::Transpose<const Matrix3> transpose() const;
+    const Eigen::Transpose<const Matrix3> transpose() const;
```
Comment (dellaert) [inline gtsam/geometry/Rot3.h]: I like it.

File: gtsam/geometry/Rot3.h [hunk 6]
```diff
@@ -435,27 +440,18 @@ namespace gtsam {
 
     /**
      * Accessor to get to component of angle representations
-     * NOTE: these are not efficient to get to multiple separate parts,
-     * you should instead use xyz() or ypr()
-     * TODO: make this more efficient
      */
-    inline double roll() const  { return ypr()(2); }
+    inline double roll() const  { return xyz()(0); }
```
Comment (dellaert) [inline gtsam/geometry/Rot3.h]: TODO of making more efficient is gone, but *is this more efficient?

File: gtsam/geometry/Point3.cpp (hunk 1 above)
Comment (varunagrawal) [reply gtsam/geometry/Point3.cpp]: Hmm I figured the printing would not affect actual behavior, but I guess for logging/metrics.

File: gtsam/geometry/Rot3.h (hunk 2 above)
Comment (varunagrawal) [reply gtsam/geometry/Rot3.h]: Good point. Sorry for overlooking that.

File: gtsam/geometry/Rot3.h (hunk 3 above)
Comment (varunagrawal) [reply gtsam/geometry/Rot3.h]: Done.

File: gtsam/geometry/Rot3.h (hunk 4 above)
Comment (varunagrawal) [reply gtsam/geometry/Rot3.h]: So be more explicit? I figured I'd make minimal changes to the code.

Comment (dellaert) [reply gtsam/geometry/Rot3.h]: More efficient. And Rot3 constructor with one argument should be explicit.

File: gtsam/geometry/Rot3.h (hunk 6 above)
Comment (varunagrawal) [reply gtsam/geometry/Rot3.h]: The TODO note said to use either `ypr` or `xyz`

```
* NOTE: these are not efficient to get to multiple separate parts,
* you should instead use xyz() or ypr()
* TODO: make this more efficient
```

I figured it was a leftover comment from a previous commit (similar to the boost TODO for rng). `ypr` was internally calling `xyz` so all I did was remove the extra layer. Internally it is still calling the `RQ` algorithm.

Comment (dellaert) [reply gtsam/geometry/Rot3.h]: So, add the TODO again :-)

**Comment by varunagrawal** [comment]
@dellaert there's a bunch of issues with the quaternion representation. I've opened #253 to track the same. Can't merge this yet. :cry:

File: gtsam/geometry/Rot3.cpp [hunk 7]
```diff
@@ -197,27 +197,15 @@ Matrix3 Rot3::LogmapDerivative(const Vector3& x)    {
...
-  // even though arithmetically, they are both atan2(0, y).
-  // Suppressing small numbers to 0.0 doesn't work since the floating point
-  // representation still causes the issue due to a delta difference.
-  // The only fix is to convert to an int so we are guaranteed the value is 0.
-  // This lambda function checks if a number is close to 0.
-  // If yes, then return the integer representation, else do nothing.
-  auto suppress = [](auto x) { return abs(x) > 1e-15 ? x : int(x); };
-
-  double x = -atan2(-suppress(A(2, 1)), suppress(A(2, 2)));
+  double x = -atan2(-A(2, 1), A(2, 2));
```
Comment (dellaert) [inline gtsam/geometry/Rot3.cpp]: Uhm, please add more justification than “I removed it’ ;-)

Comment (varunagrawal) [reply gtsam/geometry/Rot3.cpp]: This is an undo of a previous commit I had introduced in this PR. The stability is less of a problem compared to the behavior of `Eigen::Transpose`.

**Comment by varunagrawal** [comment]
`matrix().eval().transpose()` gives incorrect results which leads to test failures. Half the values end up being `1e-200` for some reason. This is something definitely to do with `Eigen::Tranpose`. The docs aren't exactly helpful and I can't figure out what I'm doing wrong. :/

**Comment by varunagrawal** [comment]
So after a LOT of investigation, there is something weird about the Eigen expression templates for transpose aka `Eigen::Transpose<const Matrix3>` when using the quaternion representation. It is  hard to figure out why since `eval()` doesn't work as it should (the transpose works except for `mat(2, 2)` which is some random `1e-10`) and I can't find anything in the docs or online to help with this.

This is clearly some issue with either how we're using the expression template or with how it is managed for quaternions. The odd thing is, if we print the value before returning, Eigen understands how to evaluate the expression correctly. The only other option is to use `eval()` in all the downstream code which would mean something like `R.transpose().eval()` which is ugly.

Any Eigen experts we can consult to figure this out? @thduynguyen?

**Comment by ProfFan** [comment]
Note that `matrix()` is a temporary and its lifetime is *the function itself*

**Comment by varunagrawal** [comment]
Even when we override it? I did `matrix().eval()` but that didn't help.

**Comment by dellaert** [comment]
@varunagrawal can't we just go back to the previous implementation? And, is quaternion one of the paths in Travis? I don't think you replied to that question.

**Comment by varunagrawal** [comment]
My last commit undid the Eigen::Transpose change while also marking some
methods as const.

Travis does have a path for Quaternion rotations which is why the template
expression behaving so strangely is a mystery to me.

File: gtsam/geometry/Rot3Q.cpp [hunk 8]
```diff
@@ -79,10 +79,14 @@ namespace gtsam {
...
-  // const Eigen::Transpose<const Matrix3> Rot3::transpose() const {
-  Matrix3 Rot3::transpose() const {
+  // TODO: Maybe use return type `const Eigen::Transpose<const Matrix3>`?
+  // It works in Rot3M but not here, because of some weird behavior
+  // due to Eigen's expression templates which needs more investigation.
+  // For example, if we use matrix(), the value returned has a 1e-10,
+  // and if we use quaternion_.toRotationMatrix(), the matrix is arbitrary.
+  // Using eval() here doesn't help, it only helps if we use it in
+  // the downstream code.
+  const Matrix3 Rot3::transpose() const {
```
Comment (dellaert) [inline gtsam/geometry/Rot3Q.cpp]: remove const

File: gtsam/geometry/Rot3Q.cpp [hunk 9]
```diff
@@ -116,7 +120,7 @@ namespace gtsam {
   }
 
   /* ************************************************************************* */
-  Matrix3 Rot3::matrix() const {return quaternion_.toRotationMatrix();}
+  const Matrix3 Rot3::matrix() const {return quaternion_.toRotationMatrix();}
```
Comment (dellaert) [inline gtsam/geometry/Rot3Q.cpp]: remove const

**Pull request merged**